drivers/stepper/gpio_stepper_controller.c

```diff
@@ -0,0 +1,231 @@
+/*
+ * SPDX-FileCopyrightText: Copyright (c) 2024 Carl Zeiss Meditec AG
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT gpio_stepper
+
+#include <zephyr/drivers/gpio.h>
+#include <zephyr/kernel.h>
+#include <zephyr/sys_clock.h>
+#include <zephyr/drivers/stepper.h>
+#include <zephyr/logging/log.h>
+#include <zephyr/sys/__assert.h>
+
+LOG_MODULE_REGISTER(gpio_stepper_motor_controller, CONFIG_STEPPER_LOG_LEVEL);
+
+#define NUMBER_OF_GPIO_PINS 4
+
+struct gpio_stepper_motor_controller_config {
+	const struct gpio_dt_spec *control_pins;
+};
+
+struct gpio_stepper_motor_controller_data {
+	const struct device *dev;
+	struct k_spinlock lock;
+	struct k_work_delayable stepper_dwork;
+	struct k_poll_signal *async_signal;
+	int32_t actual_position;
+	int32_t target_position;
+	uint32_t delay_in_us;
+};
+
+static int stepper_motor_move_with_4_gpios(const struct gpio_dt_spec *gpio_spec,
+					   int32_t step_number)
+{
+	switch (step_number) {
+	case 0:
+		LOG_DBG("Step 1010");
+		gpio_pin_configure_dt(&gpio_spec[0], GPIO_OUTPUT_ACTIVE);
+		gpio_pin_configure_dt(&gpio_spec[1], GPIO_OUTPUT_INACTIVE);
+		gpio_pin_configure_dt(&gpio_spec[2], GPIO_OUTPUT_ACTIVE);
+		gpio_pin_configure_dt(&gpio_spec[3], GPIO_OUTPUT_INACTIVE);
+		break;
+
+	case 1:
+		LOG_DBG("Step 0110");
+		gpio_pin_configure_dt(&gpio_spec[0], GPIO_OUTPUT_INACTIVE);
+		gpio_pin_configure_dt(&gpio_spec[1], GPIO_OUTPUT_ACTIVE);
+		gpio_pin_configure_dt(&gpio_spec[2], GPIO_OUTPUT_ACTIVE);
+		gpio_pin_configure_dt(&gpio_spec[3], GPIO_OUTPUT_INACTIVE);
+		break;
+
+	case 2:
+		LOG_DBG("Step 0101");
+		gpio_pin_configure_dt(&gpio_spec[0], GPIO_OUTPUT_INACTIVE);
+		gpio_pin_configure_dt(&gpio_spec[1], GPIO_OUTPUT_ACTIVE);
+		gpio_pin_configure_dt(&gpio_spec[2], GPIO_OUTPUT_INACTIVE);
+		gpio_pin_configure_dt(&gpio_spec[3], GPIO_OUTPUT_ACTIVE);
+		break;
+
+	case 3:
+		LOG_DBG("Step 1001");
+		gpio_pin_configure_dt(&gpio_spec[0], GPIO_OUTPUT_ACTIVE);
+		gpio_pin_configure_dt(&gpio_spec[1], GPIO_OUTPUT_INACTIVE);
+		gpio_pin_configure_dt(&gpio_spec[2], GPIO_OUTPUT_INACTIVE);
+		gpio_pin_configure_dt(&gpio_spec[3], GPIO_OUTPUT_ACTIVE);
+		break;
+	}
+	return 0;
+}
+
+void stepper_work_step_handler(struct k_work *work)
+{
+	struct k_work_delayable *dwork = k_work_delayable_from_work(work);
+	struct gpio_stepper_motor_controller_data *data =
+		CONTAINER_OF(dwork, struct gpio_stepper_motor_controller_data, stepper_dwork);
+	const struct gpio_stepper_motor_controller_config *config = data->dev->config;
+
+	static int32_t step_count = 1;
+
+	K_SPINLOCK(&data->lock) {
+		LOG_DBG("Moving Motor from %d to %d with delay %d ", data->actual_position,
+			data->target_position, data->delay_in_us);
+		if (data->actual_position != data->target_position) {
+			if (data->actual_position > data->target_position) {
+				stepper_motor_move_with_4_gpios(config->control_pins, step_count--);
+				step_count =
+					step_count < 0 ? (NUMBER_OF_GPIO_PINS - 1) : step_count;
+				data->actual_position--;
```

Comment:
potential integer overflow – the motor will swap the direction and count back for avery very long time

Reply:
Hi @rexut, i am going to restructure the code to stop the reversal of motion. However, generally speaking overflow of the `position` if something that cannot be avoided, or? For instance, the `TMC5041` driver has this register called actual_position which will eventually overflow if the motor keeps on running in one direction incessantly.

Reply:
Overflow of a motor owned **relative** counter (or position in your speak) must never handled as long as your counter is an **unsigned** integer! Change your focus away from absolut position handling here in this first low level stepper motor API.

A stepper motor for itself has no start or end position. It's the nature of circles!

When you talk about an absolut positioning system you have to define your null-point in this system. Where is it in yours? Think about and you will find out, that there is no common answer. Absolut positioning is a specific use case with many geometric and arithmetic boundary conditions and constraints. Believe me, you will never see such kind of mathematics inside an API that have to drive different classes of stepper motor and their power stages, incl. fault detection and event handling.

Positioning and navigation is the job of another Zephyr sub-system / API (see USB vs. USB Classes, or Ethernet PHY vs. MAC, or Sensor vs. Sensing, a.s.o.).

This Stepper Motor Driver API can do its job with absolute velocity and relative position (or steps) as API values.

Reply:
@rexut there has been a lot of discussion on this in this pull and some previous issues. Absolute stepper position is implemented out of the box by many driver ICs, usually as a register directly available for setting/getting on the chip. This driver is not trying to account for every type of BLDC; you have to start somewhere.